polaris-react/src/components/DataTable/DataTable.tsx

```diff
@@ -425,46 +433,62 @@ class DataTableInner extends PureComponent<CombinedProps, DataTableState> {
     };
   };
 
-  private stickyHeaderScrolling = (event: Event) => {
+  private focusListener = (event: Event) => {
+    this.stickyHeaderFocusStateChangeOnScroll(event);
+  }
+
+  private stickyHeaderFocusStateChangeOnScroll = (event: Event) => {
+    if (this.scrollContainer.current == null || event.target == null) {
+      return;
+    }
+
+    const target = event.target as HTMLElement;
+    const currentCell = target.parentNode as HTMLTableCellElement;
+
+    const tableScrollLeft = this.scrollContainer.current.scrollLeft;
+    const tableViewableWidth = this.scrollContainer.current.offsetWidth;
+    const tableRightEdge =  tableScrollLeft + tableViewableWidth;
+    const firstColumnWidth = this.state.columnVisibilityData[0].rightEdge;
+    const currentColumnLeftEdge = currentCell.offsetLeft;
+    const currentColumnRightEdge = currentCell.offsetLeft + currentCell.offsetWidth;
+
+    if (tableScrollLeft > currentColumnLeftEdge - firstColumnWidth) {
+      this.scrollContainer.current.scrollLeft = currentColumnLeftEdge - firstColumnWidth;
+    }
+
+    if (currentColumnRightEdge > tableRightEdge) {
+      this.scrollContainer.current.scrollLeft = currentColumnRightEdge - tableViewableWidth;
+    }
+  }
+
+  private stickyHeaderScrolling = () => {
+
     const {current: stickyTableHeadingsRow} = this.stickyTableHeadingsRow;
-    const {current: table} = this.table;
     const {current: scrollContainer} = this.scrollContainer;
-    const targetIsStickyHeader = event.target === stickyTableHeadingsRow;
 
-    if (
-      stickyTableHeadingsRow == null ||
-      scrollContainer == null ||
-      table == null
-    ) {
+    if (stickyTableHeadingsRow == null || scrollContainer == null) {
       return;
     }
 
-    const {scrollLeft: stickyScrollLeft} = stickyTableHeadingsRow;
-    const {scrollLeft: scrollContainerScrollLeft} = scrollContainer;
+    stickyTableHeadingsRow.scrollLeft = scrollContainer.scrollLeft;
 
-    if (targetIsStickyHeader) {
-      table.style.transform = `translateX(-${stickyScrollLeft}px)`;
-    } else {
-      this.stickyHeadings.map((heading) => {
-        heading.style.transform = `translateX(-${scrollContainerScrollLeft}px)`;
-      });
-    }
   };
 
-  private scrollListener = (event: Event) => {
+  private scrollListener = () => {
     debounce(() => {
       this.setState((prevState) => ({
         ...this.calculateColumnVisibilityData(prevState.condensed),
       }));
     }, 500);
 
-    this.stickyHeaderScrolling(event);
+    if (this.props.stickyHeader && this.stickyHeaderActive) {
```

Comment:
Sticky header will only fire when the table _uses_ sticky header and also sticky header is active.